Make Python op registration work with torchdeploy/multipy

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87162

See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make Python op registration work with torchdeploy/multipy

See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@d4l3k I should probably test this against the external multipy repo but I don't know how to do it. Are there instructions to follow somewhere?

d4l3k commented:
Instructions are in the readme at https://github.com/pytorch/multipy/#running-multipyruntime-build-steps-from-source

I can also help test it tomorrow if you'd like

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -145,7 +145,12 @@ OperatorEntry::AnnotatedKernelContainerIterator OperatorEntry::registerKernel(
+        // Don't complain if a PythonOpRegistrationTrampoline is overriding
+        // another PythonOpRegistrationTrampoline; you will have a
+        // registration per multipy/torchdeploy interpreter
+        !(k.front().kernel.isPythonOpRegistrationTrampoline() &&
```
is this to let multiple interpreters register a kernel for the same op, dispatch key pair?

ezyang replied:
Yes. I thought about a few ways of doing this, but the easiest was to just let the interpreters clobber each other without raising warnings.

anjali411 commented on aten/src/ATen/core/library.cpp:
```diff
@@ -218,6 +218,30 @@ Library& Library::_impl(const char* name_str, CppFunction&& f) & {
 }
+
+c10::OperatorHandle Library::_resolve(const char* name_str) const {
+  auto name = torch::jit::parseName(name_str);
```
nit -- can create a helper that does the check on namespace as the same logic is repeated in impl

ezyang replied:
yep, i'll do it after passing ci

anjali411 reviewed (approved):
It would be nice if we could verify that the function being registered for an op is not using any global state.

Also, should we add a check to ensure that all the required modules are installed in the interpreter before it accesses a function registered by another interpreter?

ezyang commented:
> It would be nice if we could verify that the function being registered for an op is not using any global state.

Yeah, it would; I don't know how to easily test this though, esp outside of a multipy environment.

> Also, should we add a check to ensure that all the required modules are installed in the interpreter before it accesses a function registered by another interpreter?

I'm not familiar enough with multipy's "startup" process to know the answer to this. @d4l3k do you know?

bdhirsh commented on c10/core/impl/HermeticPyObjectTLS.h:
```diff
@@ -0,0 +1,42 @@
+ private:
+  // This only flipped once from false to true during torchdeploy/multipy
+  // initialization, and never again.
+  static bool haveState_;
```
It looks like on startup, you'll have multiple interpreter instances starting up and setting this bool to `true`. Is that a data race / does it need to be an atomic?

bdhirsh commented on aten/src/ATen/core/library.cpp:
```diff
@@ -218,6 +218,30 @@ Library& Library::_impl(const char* name_str, CppFunction&& f) & {
+  }
+  auto r = c10::Dispatcher::singleton().findSchema(name);
+  // We always register before resolving, so this should always succeed
+  TORCH_INTERNAL_ASSERT(r.has_value());
```
nit: `findSchemaOrThrow()`, if you want to error? It probably has a better error message

ezyang replied:
Nope, don't want to error, the operator handle should be guaranteed to exist.

bdhirsh commented on aten/src/ATen/core/boxing/OperatorKernel.h:
```diff
@@ -22,6 +22,7 @@ namespace c10 {
  */
 struct TORCH_API OperatorKernel : public c10::intrusive_ptr_target {
   virtual ~OperatorKernel() = default;
+  virtual bool isPythonOpRegistrationTrampoline() const { return false; }
```
is this getting properly overridden in `PythonOpRegistrationTrampoline`? Maybe I'm just missing it.

ezyang replied:
looks like it's not lol

wconstab commented on aten/src/ATen/core/PythonOpRegistrationTrampoline.cpp:
```diff
@@ -0,0 +1,55 @@
+void PythonOpRegistrationTrampoline::operator()(const c10::OperatorHandle& op, c10::DispatchKeySet dispatch_keys, torch::jit::Stack* stack) {
+  // The simple implementation of Python op registration involves directly
+  // registering a Python callback to the dispatcher.  However, this doesn't
+  // work with torchdeploy/multipy, where there are multiple Python
```
what is multipy? (if different from torchdeploy)

ezyang replied:
https://github.com/pytorch/multipy/

bdhirsh commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -195,16 +177,25 @@ void initDispatchBindings(PyObject* module) {
+                            dispatch))));
+            // NB: this must happen AFTER the impl, because the OperatorHandle
+            // is not guaranteed to exist before we impl
+            python_registrations_[lib._resolve(name)].insert_or_assign(
```
Hmm - it seems like the key thing is:
```
  // Which interpreter should we call?  Python operator registrations operate
  // in "hermetic mode", which means that tensors passing to Python get fresh
  // Python objects (not associated with the previous interpreter), and Python
  // objects aren't preserved when the Python objects go out of scope.  This
  // means that, in fact, these calls can be serviced by ANY interpreter.  So
  // we just arbitrarily pick one and jump to it.
```

Where if I'm following correctly, the "arbitrarily pick one and jump to it" happens when the trampoline code looks up the `SafePyObj` from this map (`python_registrations_`), and it's "arbitrary" because the values in this map can be overwritten multiple times when different PyInterpreters register the same op/key/function, so it's semi-arbitrary which one registers last and leaves their function in this map.

Instead of having a second out-of-band mapping from `operator` -> `dispatch_key` -> SafePyObj` that the trampoline needs to look up, is there any reason why we can't just directly store the current function/interpreter combo (the `SafePyObj` that we want to call) directly on the trampoline? That would save us from doing a bunch of extra map lookups, and keeping a separate map around. Let me know if I'm missing something though.

ezyang replied:
Not quite. `python_registrations_` is defined in torch/csrc so it is only available from libtorch_python, but we need to make an interpreter decision in libtorch. The arbitrary interpreter is grabbed via `  static std::atomic<c10::impl::PyInterpreter*> interpreter_;` in `PythonOpRegistrationTrampoline`.

It may be possible to store the py object on the trampoline though. To do this, though, I will have to make it so that ONLY the victim interpreter actually does registrations, and everyone else whiffs them. I guess I will go implement this.

@d4l3k one thing this discussion clued me in on, is what if the victim interpreter gets unloaded before everything else is done? It seems like we will need to maintain an invariant that the first multipy interpreter always stays live.

ezyang replied:
There is also an extra subtlety, which is that if I have two interpreters racing initialization, one interpreter could attempt to use an operator registration before the first interpreter had finished initializing all of its modules. One easy fix for this is to delay actually doing the CEX for making an interpreter visible until we're done processing `import torch`, but if there are Python operator registrations from other libraries, which could possibly be lazily imported, you may have to wait arbitrarily long for the first interpreter to be ready. So here I'm not so sure how multipy can even tell when all the ops are loaded.

bdhirsh replied:
Would it help instead of using an arbitrary interpreter (depending on which one started up first), we set things up such that the "chosen" interpreter corresponds to the first interpreter to make an operator call?

Although... I'm assuming that if an interpreter is making an op call then all of it's ops have already been registered. I'm not sure if that's always the case.

And that would also have a cost, since we'd be pushing the "set up the interpreter" step from registration time to the time we invoke an operator.

ezyang replied:
If you can assume that an operator is only called after everything is loaded, this should work. But if there can be operator calls and imports interleaved, this isn't sound. In fact I don't think there is anything we can do. So I'm just going to write a nice error message.

wconstab commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1887,9 +1900,24 @@ static PyObject* THPVariable_NewWithVar(
-      var.unsafeGetTensorImpl()->set_python_dispatch(true);
+    if (c10::impl::HermeticPyObjectTLS::get_state()) {
+      // Do NOT initialize pyobj field on the tensor
+      TORCH_INTERNAL_ASSERT(
```
where does pyobj get initialized in the hermetic use case?

albanD reviewed (commented):
Looks good. No further comments beyond the existing ones.

bdhirsh reviewed (approved):
cool!

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
Unfortunately, the simple strategy (always go to one interpreter), doesn't work, because we have tensor subclasses which call into Python from TensorImpl methods (fake tensor as one notable example). We MUST send these to the same interpreter, as there is no way to run the C++ implementation hermetically.

So I need to go back to doing a hybrid strategy, which is if any of the inputs has a PyObject associated with it, we send it to the home PyInterpreter (and we only hermetic if there are no input tensors with a PyInterpreter, and no active mode.) The code is a bit more complicated, because we need the trampoline, and also all the interpreters which didn't register still need to keep the functions around in a hash table so they can service trampolined requests. Bloop.

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make Python op registration work with torchdeploy/multipy"


See strategy at PythonOpRegistrationTrampoline.cpp for the
big picture.

Along the way, I made OperatorHandle support == and hashing,
and slightly changed the low level python_dispatch impl API
to disallow empty strings for dispatch key, which had the knock
on effect of requiring us to explicitly make sure we pass in
CompositeImplicitAutograd if we would have passed in "" (I didn't apply
this to the rest of the file because I'm lazy.)

Test strategy is we delete the logic for preventing Python op
registrations in torch from being skipped in a torchdeploy context
and show CI still works.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "spurious failures"

kurtamohler commented on test/test_torch.py:
```diff
@@ -5673,36 +5673,6 @@ def test_unflatten(self):
-            raise RuntimeError("Could not 'import torch' with PYTORCH_DISABLE_LIBRARY=0") from e
-
-    # Test that warnings generated from C++ are translated to the correct type
-    def test_warn_types(self):
```
@ezyang, what was the reason for removing this test?

ezyang replied:
ugh, I think I only intended to delete the first test, not the second

kurtamohler replied:
Ah I was wondering if that was it. No worries, I can add it back in

kurtamohler replied:
Nevermind, I see that you did
